Homework/homework6/task1.py

```diff
@@ -0,0 +1,114 @@
+3. Student и Teacher имеют одинаковые по смыслу атрибуты
+(last_name, first_name) - избавиться от дублирования с помощью наследования
+4.
+Teacher
+Атрибут:
+    homework_done - структура с интерфейсом как в словаря, сюда поподают все
+    HomeworkResult после успешного прохождения check_homework
+    (нужно гаранитровать остутствие повторяющихся результатов по каждому
+    заданию), группировать по экземплярам Homework.
+    Общий для всех учителей. Вариант ипользования смотри в блоке if __main__...
+Методы:
+    check_homework - принимает экземпляр HomeworkResult и возвращает True если
+    ответ студента больше 5 символов, так же при успешной проверке добавить в
+    homework_done.
+    Если меньше 5 символов - никуда не добавлять и вернуть False.
+    reset_results - если передать экземпряр Homework - удаляет только
+    результаты этого задания из homework_done, если ничего не передавать,
+    то полностью обнулит homework_done.
+PEP8 соблюдать строго.
+Всем перечисленным выше атрибутам и методам классов сохранить названия.
+К названием остальных переменных, классов и тд. подходить ответственно -
+давать логичные подходящие имена.
+"""
+import datetime
+from collections import defaultdict
+
+
+class DeadlineError(Exception):
+    """Error raises when homework's deadline passed."""
+
+
+class Homework:
+    def __init__(self, text: str, deadline: int):
+        self.text = text
+        self.deadline = datetime.timedelta(days=deadline)
+        self.created = datetime.datetime.now()
+
+    def is_active(self) -> bool:
+        return self.created + self.deadline > datetime.datetime.now()
+
+
+class HomeworkResult:
+    def __init__(self, author, homework: Homework, solution: str):
+        if isinstance(homework, Homework):
+            self.homework = homework
+        else:
+            raise TypeError("You gave not a Homework object")
+        self.solution = solution
+        self.author = author
+        self.created = datetime.datetime.now()
+
+
+class Person:
+    def __init__(self, first_name: str, last_name: str):
+        self.first_name = first_name
+        self.last_name = last_name
+
+
+class Student(Person):
+    def do_homework(self, homework: Homework, solution: str) -> HomeworkResult:
+        if not isinstance(homework, Homework):
+            raise TypeError("You gave not a Homework object")
+        if homework.is_active():
+            return HomeworkResult(self, homework, solution)
+        raise DeadlineError("You are late")
+
+
+class Teacher(Person):
+
+    homework_done = defaultdict(list)
+
+    def check_homework(self, homework_result: HomeworkResult) -> bool:
+        if not isinstance(homework_result, HomeworkResult):
+            raise TypeError("You gave not a HomeworkResult object")
+        homework = homework_result.homework
+        if (
+            len(homework_result.solution) > 5
+            and homework_result not in self.homework_done[homework]
```

Comment:
How else we can maintain collection of unique elements?

Reply:
Using `set`